Realm/Tests/NotificationTests.m

```diff
@@ -1256,4 +1270,129 @@ - (void)testDiffedUpdatesOnlyNotifyForPropertiesWhichActuallyChanged {
+    id ex = [self expectationWithDescription:@"change notification"];
+    RLMNotificationToken *token = [_obj addNotificationBlock:^(BOOL deleted, NSArray *changes, NSError *error) {
+        XCTAssertFalse(deleted);
+        XCTAssertNil(error);
+        XCTAssertEqual(changes.count, 1U);
+        RLMPropertyChange *prop = changes[0];
+        XCTAssertEqualObjects(prop.name, @"boolCol");
+        [ex fulfill];
+    } keyPaths:@[@"boolCol"]];
+
+    [_obj.realm beginWriteTransaction];
+    XCTAssertNotEqual(_obj.boolCol, [_values[@"boolCol"] boolValue]);
+    _obj.boolCol = _values[@"boolCol"];
+    [_obj.realm commitWriteTransaction];
+    [self waitForExpectationsWithTimeout:0.1 handler:nil];
+}
+
+-(void) testModifyUnobservedKeyPathLocally {
+    XCTestExpectation *ex = [self expectationWithDescription:@"no change notification"];
+    ex.inverted = @YES;
+    RLMNotificationToken *token = [_obj addNotificationBlock:^(BOOL deleted, NSArray *changes, NSError *error) {
+        [ex fulfill];
+    } keyPaths:@[@"boolCol"]];
+
+    [_obj.realm beginWriteTransaction];
+    _obj.intCol = _obj.intCol + _obj.intCol;
+    [_obj.realm commitWriteTransaction];
+    [self waitForExpectationsWithTimeout:0.1 handler:nil];
+}
+
+-(void) testModifyObservedKeyPathRemotely {
+    id ex = [self expectationWithDescription:@"change notification"];
+    RLMNotificationToken *token = [_obj addNotificationBlock:^(BOOL deleted, NSArray *changes, NSError *error) {
+        XCTAssertFalse(deleted);
+        XCTAssertNil(error);
+        XCTAssertEqual(changes.count, 1U);
+        RLMPropertyChange *prop = changes[0];
+        XCTAssertEqualObjects(prop.name, @"boolCol");
+        [ex fulfill];
+    } keyPaths:@[@"boolCol"]];
+
+    [self dispatchAsync:^{
+        RLMRealm *realm = [RLMRealm defaultRealm];
+        AllTypesObject *obj = [[AllTypesObject allObjectsInRealm:realm] firstObject];
+        [realm beginWriteTransaction];
+        XCTAssertNotEqual(obj.boolCol, [_values[@"boolCol"] boolValue]);
+        obj.boolCol = _values[@"boolCol"];
+        [realm commitWriteTransaction];
+    }];
+    [self waitForExpectationsWithTimeout:0.1 handler:nil];
+}
+
+-(void) testModifyUnobservedKeyPathRemotely {
+    XCTestExpectation *ex = [self expectationWithDescription:@"no change notification"];
+    ex.inverted = @YES;
+
+    RLMNotificationToken *token = [_obj addNotificationBlock:^(BOOL deleted, NSArray *changes, NSError *error) {
+        [ex fulfill];
+    } keyPaths:@[@"boolCol"]];
+
+    [self dispatchAsync:^{
+        RLMRealm *realm = [RLMRealm defaultRealm];
+        AllTypesObject *obj = [[AllTypesObject allObjectsInRealm:realm] firstObject];
+        [realm beginWriteTransaction];
+        obj.intCol = obj.intCol + obj.intCol;
+        [realm commitWriteTransaction];
+    }];
+    [self waitForExpectationsWithTimeout:0.1 handler:nil];
+}
+
+-(void) testModifyObservedKeyPathArrayProperty {
+    id ex = [self expectationWithDescription:@"change notification"];
```

Comment:
Is there a reason for sometimes using `XCTestExpectation *ex` and sometimes `id ex`?

Reply:
Not a good one - I originally defined all expectations as id. But then needed to define them as XCTTestExpectation in order to invert some.
I'll make them uniform.

Reply:
Should be resolved [here](https://github.com/realm/realm-cocoa/pull/7249/commits/9465b5d3b85471e46621fa77d7c62c313ad29127)